accounts-db/src/accounts.rs

```diff
@@ -385,15 +386,26 @@ impl Accounts {
                             .load_with_fixed_root(ancestors, key)
                             .map(|(mut account, _)| {
                                 if message.is_writable(i) {
-                                    let rent_due = rent_collector
-                                        .collect_from_existing_account(
-                                            key,
-                                            &mut account,
-                                            self.accounts_db.filler_account_suffix.as_ref(),
-                                            set_exempt_rent_epoch_max,
-                                        )
-                                        .rent_amount;
-                                    (account.data().len(), account, rent_due)
+                                    if should_collect_rent {
+                                        let rent_due = rent_collector
+                                            .collect_from_existing_account(
+                                                key,
+                                                &mut account,
+                                                self.accounts_db.filler_account_suffix.as_ref(),
+                                                set_exempt_rent_epoch_max,
+                                            )
+                                            .rent_amount;
+
+                                        (account.data().len(), account, rent_due)
+                                    } else {
+                                        // All rent-exempt accounts need to have their rent_epoch set to u64::max. This is done inside
+                                        // `collect_from_existing_account` when we collect rent. When rent collection is disabled,
+                                        // we should also enforce that the rent_epoch on these accounts is set to u64::MAX too.
+                                        if set_exempt_rent_epoch_max {
+                                            account.set_rent_epoch(u64::MAX);
+                                        }
```

Comment:
We are not collecting rent means that the account is rent-exempted. It looks like that runtime will require that for rent-exempted accounts to have its rent_epoch set to u64:MAX, at least as indicated in test_transaction_result_does_not_affect_bankhash. 

That's why I added https://github.com/solana-labs/solana/pull/33945/commits/e1aee986edfee52ecddc4a2d378ba68919cb074d after seeing that test failure.

Reply:
> When we create new accounts now, they must be rent exempt. Do we set the rent epoch to `RENT_EPOCH_RENT_EXEMPT` at that moment

I think and hope this is what we're doing. Maybe in some test case this isn't happening?

Reply:
> Looking at test_transaction_result_does_not_affect_bankhash(), it's not clear to me how the rent epoch would be an issue. Do you happen to have a log of the failure? Maybe the test is what needs to be updated.

All of our tests that use genesis to populate initial accounts violate the assumption that new accounts are rent exempted. 
https://github.com/solana-labs/solana/blob/d93280c28731310f4d14d1a1f50b5e81a8bc4ab6/runtime/src/genesis_utils.rs#L249

I am not sure if it is we want to update `genesis_utils` and all its affected tests. 

> If we have a bug, or hole, and a rent-paying account is created, then we probably still want to know it happened. Even if we're not collecting rent, this account should still be considered rent-paying.
Yes, we want to know. There are other ways to know it. I think TotalAccountsStats will tell us. 
https://github.com/solana-labs/solana/blob/ffa462dd3aa6e6dca37afaa45150ddc3bf36ac03/runtime/src/bank.rs#L8343

Reply:
The test passed because of [the set_rent epoch ](https://github.com/solana-labs/solana/pull/33945/commits/e1aee986edfee52ecddc4a2d378ba68919cb074d). 

If you remove this commit, I think the test will fail. 

The fix for the test without the above commit will be to not check the blockhash.

Reply:
I pushed a commit to implement the behavior that you suggested. @brooksprumo Let me know wdyt?

Reply:
I pushed the commit to update with the above code. 

Note that with this feature, the `rent_epoch` field on the rent paying account behavior will change slightly. Since account `rent_epoch` are currently exposed to user program at runtime, any user program that depends on the old behavior of the `rent_epoch` update might break. Since rent paying account will no longer exists, we believe this change won't impact any user program.